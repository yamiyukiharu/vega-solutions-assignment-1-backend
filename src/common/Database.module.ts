import { Module } from "@nestjs/common";
import { TypeOrmModule } from "@nestjs/typeorm";
import { AppConfigModule } from "./AppConfig.module";
import { ConfigService } from "@nestjs/config";
import { Transaction } from "src/transaction/models/Transaction.model";

@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      imports: [AppConfigModule],
      useFactory: (configService: ConfigService) => ({
        type: 'mongodb',
        url: configService.get('DB_URL'),
        entities: [Transaction],
        synchronize: true,
        useNewUrlParser: true,
        logging: true,
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [TypeOrmModule]
})
export class DatabaseModule { }