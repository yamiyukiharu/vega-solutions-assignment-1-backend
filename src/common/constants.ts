export const REPORTS_QUEUE = 'transactions-report';

export const RECORD_QUEUE = 'transactions-record';