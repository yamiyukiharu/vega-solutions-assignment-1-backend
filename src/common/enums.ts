export enum Protocol {
  UNISWAPV3 = 'uniswapv3',
}

export enum Pool {
  ETH_USDC = 'usdc_eth',
}

export enum ReportStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  FAILED = 'failed',
}