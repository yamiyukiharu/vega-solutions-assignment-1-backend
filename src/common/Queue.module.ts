
import { BullModule } from '@nestjs/bull';
import { Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { AppConfigModule } from './AppConfig.module';

@Module({
  imports: [
    BullModule.registerQueueAsync({
      imports: [AppConfigModule],
      name: 'transactions-report',
      useFactory: async (configService: ConfigService) => ({
        redis: {
          host: configService.get('QUEUE_HOST'),
          port: configService.get('QUEUE_PORT'),
        },
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [BullModule],
})
export class QueueModule { }
