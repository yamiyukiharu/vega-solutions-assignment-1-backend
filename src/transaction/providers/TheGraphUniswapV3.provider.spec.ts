import { Test, TestingModule } from '@nestjs/testing';
import { HttpService } from '@nestjs/axios';
import { TheGraphUniswapV3Provider, TheGraphResponse } from './TheGraphUniswapV3.provider';
import { GetTransactionOptions } from './ITransaction.provider';
import { Pool } from 'src/common/enums';
import BigNumber from 'bignumber.js';

describe('TheGraphUniswapV3Provider', () => {
  let provider: TheGraphUniswapV3Provider;
  let httpService: HttpService;
  const mockAxiosPost = jest.fn();

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        TheGraphUniswapV3Provider,
        {
          provide: HttpService,
          useValue: {
            axiosRef: {
              post: mockAxiosPost,
            },
          },
        },
      ],
    }).compile();

    provider = module.get<TheGraphUniswapV3Provider>(TheGraphUniswapV3Provider);
    httpService = module.get<HttpService>(HttpService);
  });

  describe('getTransactions', () => {
    const mockResponse: TheGraphResponse = {
      data: {
        swaps: [
          {
            amount0: '100',
            amount1: '200',
            transaction: {
              id: '0x123',
              blockNumber: '123',
              gasUsed: '100000',
              gasPrice: '1000000000',
              timestamp: '1628000000',
            },
          },
        ],
      },
    };

    const mockOptions: GetTransactionOptions = {
      pool: Pool.ETH_USDC,
      page: 0,
      limit: 100,
      startTimestamp: 1628000000,
      endTimestamp: 1629000000,
      sort: 'asc',
    };

    it('should return an array of transactions', async () => {
      jest.spyOn(httpService.axiosRef, 'post').mockResolvedValueOnce({ data: mockResponse });

      const result = await provider.getTransactions(mockOptions);

      expect(mockAxiosPost.mock.calls[0]).toMatchSnapshot();

      expect(result).toEqual([
        {
          id: '0x123',
          timestamp: 1628000000,
          blockNumber: 123,
          fee: '100000000000000',
          swapPrice: '0.5',
        },
      ]);
    });
  });
});