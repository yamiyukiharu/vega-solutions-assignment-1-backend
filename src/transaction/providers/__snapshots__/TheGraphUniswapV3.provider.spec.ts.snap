// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheGraphUniswapV3Provider getTransactions should return an array of transactions 1`] = `
[
  "https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v3",
  {
    "query": "
    query {
      swaps(
        first: 100,
        skip: 0,
        orderBy: timestamp,
        orderDirection: asc,
        where: { pool: "0x88e6a0c2ddd26feeb64f039a2c41296fcb3f5640", timestamp_gte: "1628000000", timestamp_lte: "1629000000" },
      ) {
        amount0
        amount1
        transaction {
          id
          blockNumber
          gasUsed
          gasPrice
          timestamp
        }
      }
    }
    ",
  },
]
`;
